src/contract_client.rs

```diff
@@ -668,6 +677,193 @@ impl ModuleDeployHandle {
+/// are given as such that the meaning of the decoded reject reason haven't been
+/// altered by the smart contract logic.
+pub fn decode_smart_contract_revert(
+    return_value: Option<ReturnValue>,
+    reject_reason: RejectReason,
+    schema: &Option<VersionedModuleSchema>,
+    receive_name: OwnedReceiveName,
+    contract_name: &OwnedContractName,
+) -> Result<Option<String>, RejectReasonDecodeError> {
+    match reject_reason {
+        RejectReason::RejectedReceive {
+            reject_reason: reject_reason_code,
+            contract_address: _,
+            receive_name: _,
+            parameter: _,
+        } => {
+            // Step 1: Try to decode the `reject_reason` using the `concordium-std`
+            // error codes.
+            if let Some(decoded_error) = decode_concordium_std_error(reject_reason_code) {
+                return Ok(Some(decoded_error));
+            }
+
+            // Step 2: Try to decode the `reject_reason` using the `error_schema` and the
+            // `return_value`.
+            if let Some(schema) = schema {
+                if let (Some(error_schema), Some(return_value)) = (
+                    get_error_schema(schema, receive_name, contract_name),
+                    return_value,
+                ) {
+                    let mut cursor = Cursor::new(return_value.value);
+
+                    match error_schema.to_json(&mut cursor) {
+                        Ok(serde_json::Value::Object(obj)) => {
+                            if let Some(key) = obj.keys().next() {
+                                return Ok(Some(key.to_string()));
+                            }
+                            // This error can not happen if a valid `error_schema` is provided.
+                            return Err(RejectReasonDecodeError::MissingErrorVariant);
+                        }
+                        Ok(_) => {
+                            // This error can not happen if a valid `error_schema` is provided.
+                            return Err(RejectReasonDecodeError::NoMatchingErrorVariant);
+                        }
+                        // This error can not happen if a valid `error_schema` is provided.
+                        Err(e) => {
+                            return Err(e.into());
+                        }
+                    }
+                }
+            }
+
+            // If no `error_schema` and/or `return_value` is provided, the `reject_reason`
+            // can not be decoded.
+            Ok(None)
+        }
+        // If the error is NOT caused by a smart contract logical revert, the
+        // `reject_reason` is already a human-readable error. No
+        // further decoding of the error is needed. An example of
+        // such a transaction is the error variant "OutOfEnergy".
+        _ => Ok(None),
+    }
+}
+
 impl<Type> ContractClient<Type> {
     /// Construct a [`ContractClient`] by looking up metadata from the chain.
```

Comment:
***Suggestion:** Include the information that it will attempt to get the embedded schema on chain as well